batcher/sweep_batcher.go

```diff
@@ -0,0 +1,286 @@
+package sweepbatcher
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+const (
+	// DefaultMaxTimeoutDistance is the default maximum timeout distance
+	// of sweeps that can appear in the same batch.
+	DefaultMaxTimeoutDistance = 288
+)
+
+type BatcherStore interface {
+	// FetchUnconfirmedSweepBatches fetches all the batches from the database that are
+	// not in a confirmed state.
+	FetchUnconfirmedSweepBatches(ctx context.Context) ([]*loopdb.Batch, error)
+
+	// UpsertSweepBatch inserts a batch into the database, or updates an existing batch
+	// if it already exists.
+	UpsertSweepBatch(ctx context.Context, batch *loopdb.Batch) error
+
+	// ConfirmBatch confirms a batch by setting its state to confirmed.
+	ConfirmBatch(ctx context.Context, id []byte) error
+}
+
+type MuSig2SignSweep func(ctx context.Context,
+	protocolVersion loopdb.ProtocolVersion, swapHash lntypes.Hash,
+	paymentAddr [32]byte, nonce []byte, sweepTxPsbt []byte,
+	prevoutMap map[wire.OutPoint]*wire.TxOut) (
+	[]byte, []byte, error)
+
+// Batcher is a system that is responsible for accepting sweep requests and
+// placing them in appropriate batches. It will spin up new batches as needed.
+type Batcher struct {
+	// batches is a map of batch IDs to the currently active batches.
+	batches map[lntypes.Hash]*Batch
+
+	// SweepReqs is a channel where sweep requests are received.
+	SweepReqs chan *Sweep
+
+	// errChan is a channel where errors are received.
+	errChan chan error
+
+	// completeChan is a channel where batches signal their completion by
+	// providing their ID.
+	completeChan chan lntypes.Hash
+
+	// wallet is the wallet kit client that is used by batches.
+	wallet lndclient.WalletKitClient
+
+	// chainNotifier is the chain notifier client that is used by batches.
+	chainNotifier lndclient.ChainNotifierClient
+
+	// signerClient is the signer client that is used by batches.
+	signerClient lndclient.SignerClient
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
```

Comment:
nit: could use a functional options pattern or a config here to help readability.

Reply:
I followed this approach for the `NewBatch` method. The `NewBatcher` method only has 4 arguments so I think creating a config would be a bit overkill